docs/design/copy-tracking.md

```diff
@@ -0,0 +1,286 @@
+# Copy Tracking and Tracing Design
+
+Authors: [Daniel Ploch](mailto:dploch@google.com)
+
+**Summary:** This Document documents an approach to tracking and detecting copy
+information in jj repos, in a way that is compatible with both Git's detection
+model and with custom backends that have more complicated tracking of copy
+information. This design affects the output of diff commands as well as the
+results of rebasing across remote copies.
+
+## Objective
+
+Implement extensible APIs for recording and retrieving copy info for the
+purposes of diffing and rebasing across renames and copies more accurately.
+This should be performant both for Git, which synthesizes copy info on the fly
+between arbitrary trees, and for custom extensions which may explicitly record
+and re-serve copy info over arbitrarily large commit ranges.
+
+The APIs should be defined in a way that makes it easy for custom backends to
+ignore copy info entirely until they are ready to implement it.
+
+## Interface Design
+
+### Read API
+
+Copy information will be served both by a new Backend trait method described
+below, as well as a new field on Commit objects for backends that support copy
+tracking:
+
+```rust
+/// An individual copy source.
+pub struct CopySource {
+    /// The source path a target was copied from.
+    ///
+    /// It is not required that the source path is different than the target
+    /// path. A custom backend may choose to represent 'rollbacks' as copies
+    /// from a file unto itself, from a specific prior commit.
+    path: RepoPathBuf,
+    file: FileId,
+    /// The source commit the target was copied from. If not specified, then the
+    /// parent of the target commit is the source commit. Backends may use this
+    /// field to implement 'integration' logic, where a source may be
+    /// periodically merged into a target, similar to a branch, but the
+    /// branching occurs at the file level rather than the repository level. It
+    /// also follows naturally that any copy source targeted to a specific
+    /// commit should avoid copy propagation on rebasing, which is desirable
+    /// for 'fork' style copies.
+    commit: Option<CommitId>,
```

Comment:
If we don't support this upstream then any extension that does support it would need to rewrite 'diff', 'rebase', everything etc. that reads the copy source contents.  I'll leave this as-is for now, can iterate further in other PRs.